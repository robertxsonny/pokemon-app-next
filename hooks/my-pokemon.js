import { useCallback, useEffect } from "react";
import { useState } from "react";

export const useMyPokemons = () => {
  const [caughtPokemons, setCaughtPokemons] = useState(() => {
    if (typeof window === "undefined") {
      return {}
    }

    const caughtPokemonsStr = localStorage.getItem('caught-pokemons');
    return caughtPokemonsStr ? JSON.parse(caughtPokemonsStr) : {};
  });

  useEffect(() => {
    localStorage.setItem('caught-pokemons', JSON.stringify(caughtPokemons));
  }, [caughtPokemons])

  const caughtPokemonNicknames = Object.keys(caughtPokemons);

  const hasPokemonWithNickname = useCallback((nickname) => caughtPokemonNicknames.includes(nickname), [caughtPokemonNicknames]);

  const getPokemonName = useCallback((nickname) => caughtPokemons[nickname], [caughtPokemons]);

  const getNumOfCollected = useCallback((name) => Object.values(caughtPokemons).filter(v => v === name).length, [caughtPokemons]);

  const catchPokemon = useCallback((name, nickname) => {
    if (!caughtPokemons[nickname]) {
      const newCaughtPokemons = {
        ...caughtPokemons,
        [nickname]: name
      }

      setCaughtPokemons(newCaughtPokemons);
    }
  }, [caughtPokemons]);

  const releasePokemon = useCallback((nicknameToRelease) => {
    const newCaughtPokemons = Object.keys(caughtPokemons).reduce((pokemons, nickname) => {
      return nickname === nicknameToRelease ? pokemons : { ...pokemons, [nickname]: caughtPokemons[nickname] }
    }, {})

    setCaughtPokemons(newCaughtPokemons);
  }, [caughtPokemons]);

  return { caughtPokemons, hasPokemonWithNickname, getNumOfCollected, getPokemonName, catchPokemon, releasePokemon };
}
