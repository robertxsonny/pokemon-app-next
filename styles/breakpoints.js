const generateBreakpoint = (bp) => ({ max: `@media (max-width: ${bp - 1}px)`, min: `@media (min-width: ${bp}px)` })

export const xs = generateBreakpoint(576);
export const sm = generateBreakpoint(768);
export const md = generateBreakpoint(992);
export const lg = generateBreakpoint(1200)