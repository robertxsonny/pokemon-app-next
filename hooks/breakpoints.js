import { useLayoutEffect, useState } from 'react';
import { xs, sm, md, lg, xl } from '../styles/breakpoints';

export const useBreakpoints = (breakpoint, layoutDir) => {
  const [width, setWidth] = useState(undefined);

  useLayoutEffect(() => {
    const handleResize = () => {
      setWidth(window.innerWidth);
    }

    window.addEventListener('resize', handleResize);
    handleResize();
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  const breakpoints = { xs, sm, md, lg, xl }
  
  if (!breakpoints.hasOwnProperty(breakpoint)) {
    return false;
  }

  const { min, max } = breakpoints[breakpoint];

  if (layoutDir === 'up') { return width >= min; }
  if (layoutDir === 'down') { return width <= max; }
  return (width >= min && width <= max);
}