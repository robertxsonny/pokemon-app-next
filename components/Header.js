import styled from '@emotion/styled';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { md, sm } from '../styles/breakpoints';

const HeaderWrapper = styled.header(
  {
    height: 50,
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    zIndex: 99,
    display: 'flex',
    alignItems: 'center',
    backgroundColor: 'whitesmoke',
    boxShadow: '0 6px 6px rgba(0, 0, 0, 0.078)'
  },
  ({ hideOnSm }) => hideOnSm ? { [sm.max]: { display: 'none' } } : {}
)

const HeaderContainer = styled.div({
  width: '100%',
  maxWidth: '800px',
  margin: '0 auto',
  padding: '0 24px',
  boxSizing: 'border-box',
  fontSize: 20
})

const LargeOnlyHeaderContainer = styled(HeaderContainer)({
  [sm.max]: {
    display: 'none'
  },
  [md.min]: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center'
  }
})

const SmallOnlyHeaderContainer = styled(HeaderContainer)({
  [md.min]: {
    display: 'none'
  }
})

const LargeOnlyTitle = styled.div({
  'h1': {
    marginBottom: 0,
    fontSize: 24,
    textAlign: 'left',
  },
  [sm.max]: {
    display: 'none'
  }
})

const HomeLink = styled.span({
  textDecoration: 'none',
  cursor: 'pointer',
  '&:hover': {
    textDecoration: 'underline'
  }
})

const RightLinkWrapper = styled.div({ display: 'flex' });

const MenuLink = styled(HomeLink)(
  { fontSize: 14, marginLeft: 12 },
  ({ current }) => (current ? {
    cursor: 'default',
    fontWeight: 500,
    '&:hover': {
      textDecoration: 'none'
    }
  } : {})
);

const Header = ({ title, hideOnSm }) => {
  const router = useRouter();

  const { pathname } = router;

  const renderRightMenu = (path, title) => {
    return pathname === path ? <MenuLink current>{title}</MenuLink> : <Link href={path} passHref><MenuLink>{title}</MenuLink></Link>
  }

  return (
    <>
      <HeaderWrapper hideOnSm={hideOnSm}>
        <LargeOnlyHeaderContainer>
          <Link href="/" passHref>
            <HomeLink>Pokemon List App</HomeLink>
          </Link>
          <RightLinkWrapper>
            {renderRightMenu('/pokemons', 'All Pokemons')}
            {renderRightMenu('/my-pokemon', 'My collections')}
          </RightLinkWrapper>
        </LargeOnlyHeaderContainer>
        {!hideOnSm && <SmallOnlyHeaderContainer>{title}</SmallOnlyHeaderContainer>}
      </HeaderWrapper>
      {!hideOnSm && <LargeOnlyTitle><h1>{title}</h1></LargeOnlyTitle>}
    </>
  )
}

export default Header;