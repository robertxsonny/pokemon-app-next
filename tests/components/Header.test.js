import { render } from '@testing-library/react';
import { useRouter } from 'next/router';

import Header from '../../components/Header';
import { useBreakpoints } from '../../hooks/breakpoints';

jest
  .mock('../../hooks/breakpoints', () => ({
    useBreakpoints: jest.fn()
  }))
  .mock('next/router', () => ({
    useRouter: jest.fn(() => ({ pathname: '/' }))
  }))

describe('Header component', () => {
  test('Should render all components in large screen', () => {
    useBreakpoints.mockReturnValueOnce(false);
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  })

  test('Should render header menu in active state if the current page matches with the link', () => {
    useBreakpoints.mockReturnValueOnce(false);

    useRouter.mockReturnValueOnce({ pathname: '/pokemons' });
    const { container, rerender } = render(<Header />);
    expect(container).toMatchSnapshot();

    useRouter.mockReturnValueOnce({ pathname: '/my-pokemon' });
    rerender(<Header />);
    expect(container).toMatchSnapshot();
  })

  test('Should render title and back button in small screen', () => {
    useBreakpoints.mockReturnValueOnce(true);
    const { container } = render(<Header mobileTitle="Mobile" />);
    expect(container).toMatchSnapshot();
  })

  test('Should not render header in small screen if is set to hide on small screen', () => {
    useBreakpoints.mockReturnValueOnce(true);
    const { container } = render(<Header mobileTitle="Mobile" hideOnSm />);
    expect(container).toMatchSnapshot();
  })
});