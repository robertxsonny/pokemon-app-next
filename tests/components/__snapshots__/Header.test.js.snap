// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Should not render header in small screen if is set to hide on small screen 1`] = `<div />`;

exports[`Header component Should render all components in large screen 1`] = `
<div>
  <header
    class="css-r752xb-HeaderWrapper e1hd5i84"
  >
    <div
      class="css-19mqaid-HeaderContainer e1hd5i83"
    >
      <span
        class="css-o7n0c5-HomeLink e1hd5i82"
        href="/"
      >
        Pokemon List App
      </span>
      <div
        class="css-18c0i80-RightLinkWrapper e1hd5i81"
      >
        <span
          class="css-16f59f8-HomeLink-MenuLink e1hd5i80"
          href="/pokemons"
        >
          All Pokemons
        </span>
        <span
          class="css-16f59f8-HomeLink-MenuLink e1hd5i80"
          href="/my-pokemon"
        >
          My Collections
        </span>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component Should render header menu in active state if the current page matches with the link 1`] = `
<div>
  <header
    class="css-r752xb-HeaderWrapper e1hd5i84"
  >
    <div
      class="css-19mqaid-HeaderContainer e1hd5i83"
    >
      <span
        class="css-o7n0c5-HomeLink e1hd5i82"
        href="/"
      >
        Pokemon List App
      </span>
      <div
        class="css-18c0i80-RightLinkWrapper e1hd5i81"
      >
        <span
          class="css-or14iv-HomeLink-MenuLink e1hd5i80"
        >
          All Pokemons
        </span>
        <span
          class="css-16f59f8-HomeLink-MenuLink e1hd5i80"
          href="/my-pokemon"
        >
          My Collections
        </span>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component Should render header menu in active state if the current page matches with the link 2`] = `
<div>
  <header
    class="css-r752xb-HeaderWrapper e1hd5i84"
  >
    <div
      class="css-19mqaid-HeaderContainer e1hd5i83"
    >
      <span
        class="css-o7n0c5-HomeLink e1hd5i82"
        href="/"
      >
        Pokemon List App
      </span>
      <div
        class="css-18c0i80-RightLinkWrapper e1hd5i81"
      >
        <span
          class="css-16f59f8-HomeLink-MenuLink e1hd5i80"
          href="/pokemons"
        >
          All Pokemons
        </span>
        <span
          class="css-or14iv-HomeLink-MenuLink e1hd5i80"
        >
          My Collections
        </span>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component Should render title and back button in small screen 1`] = `
<div>
  <header
    class="css-r752xb-HeaderWrapper e1hd5i84"
  >
    <div
      class="css-19mqaid-HeaderContainer e1hd5i83"
    >
      <div>
        Mobile
      </div>
    </div>
  </header>
</div>
`;
